src/env/double_write.rs

```diff
@@ -0,0 +1,722 @@
+use crossbeam::channel::unbounded;
+use crossbeam::channel::Sender;
+use fail::fail_point;
+use log::{info, warn};
+use std::fs;
+use std::io::{Read, Result as IoResult, Seek, SeekFrom, Write};
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::sync::RwLock;
+use std::thread;
+
+use crate::env::{FileSystem, Handle, Permission, WriteExt};
+use futures::channel::oneshot;
+use futures::executor::block_on;
+use futures::select;
+
+use either::Either;
+
+type Callback<T> = Box<dyn FnOnce(IoResult<Option<T>>) + Send>;
+
+#[derive(PartialEq)]
+enum Task {
+    Create(PathBuf),
+    Open {
+        path: PathBuf,
+        perm: Permission,
+    },
+    Delete(PathBuf),
+    Rename {
+        src_path: PathBuf,
+        dst_path: PathBuf,
+    },
+    Stop,
+}
+
+#[derive(Default)]
+struct Files {
+    prefix: PathBuf,
+    append_file: Vec<FileName>,
+    rewrite_file: Vec<FileName>,
+    recycled_file: Vec<FileName>,
+}
+
+fn replace_path(path: &Path, from: &Path, to: &Path) -> PathBuf {
+    if let Ok(file) = path.strip_prefix(from) {
+        to.to_path_buf().join(file)
+    } else {
+        panic!("Invalid path: {:?}", path);
+    }
+}
+
+pub struct HedgedFileSystem<F: FileSystem> {
+    base: Arc<F>,
+
+    path1: PathBuf,
+    path2: PathBuf,
+    disk1: Sender<(Task, Callback<F::Handle>)>,
+    disk2: Sender<(Task, Callback<F::Handle>)>,
+
+    handle1: Option<thread::JoinHandle<()>>,
+    handle2: Option<thread::JoinHandle<()>>,
+}
+
+// TODO: read both dir at recovery, maybe no need? cause operations are to both
+// disks TODO: consider encryption
+
+impl<F: FileSystem> HedgedFileSystem<F> {
+    pub fn new(base: Arc<F>, path1: PathBuf, path2: PathBuf) -> Self {
+        let (tx1, rx1) = unbounded::<(Task, Callback<F::Handle>)>();
+        let (tx2, rx2) = unbounded::<(Task, Callback<F::Handle>)>();
+        let fs1 = base.clone();
+        let handle1 = thread::spawn(|| {
+            for (task, cb) in rx1 {
+                if task == Task::Stop {
+                    break;
+                }
+                fail_point!("double_write::thread1");
+                let res = Self::process(&fs1, task);
+                cb(res);
+            }
+        });
+        let fs2 = base.clone();
+        let handle2 = thread::spawn(|| {
+            for (task, cb) in rx2 {
+                if task == Task::Stop {
+                    break;
+                }
+                let res = Self::process(&fs2, task);
+                cb(res);
+            }
+        });
+        Self {
+            base,
+            path1,
+            path2,
+            disk1: tx1,
+            disk2: tx2,
+            handle1: Some(handle1),
+            handle2: Some(handle2),
+        }
+    }
+
+    pub fn bootstrap(&self) -> Result<()> {
+        // catch up diff
+        let files1 = self.get_files(&self.path1)?;
+        let files2 = self.get_files(&self.path2)?;
```

Comment:
In the case of `enable-log-recycle` is enabled, IIUC,  file names are reused. Meaning the same file name might hold different raft log content over time. Is it possible that for the same file name `f`,  the file on disk1 and the one on disk2 hold different contents?

Reply:
Impossible, log-recycle doesn't reuse file name.

Reply:
oh my bad, my understanding was wrong. How does log recycling work?